### ISSUE TITLE
libltc decoder returning zero

### ISSUE THREAD

**Issue by zhiguang1** [x42/libltc#26]
I'm reading in realtime LTC from a timecode generator in through PortAudio in order to record it.  Then I pass samples to the decoder:

////
data.decoder = ltc_decoder_create(data.apv, 1024);
////

in the callback function:

	ltc_decoder_write_float(data->decoder, &data->recordedSamples[6][0], framesToCalc, data->frameIndex);

	printf("queue %d",ltc_decoder_queue_length(data->decoder));

	while (ltc_decoder_read(data->decoder, &data->frame)) {
		SMPTETimecode stime;

		ltc_frame_to_time(&stime, &data->frame.ltc, 1);
... etc etc

i've tried changing the window size, apv, queue size, etc, all returning zero.  it seems that the maximum and minimum are being read, however.

**Comment by x42** [comment]
That looks correct.  Also "returning zero" means no error, I guess you're not seeing any decoded Timecode values.

Is  `&data->recordedSamples[6][0]` an array of non-interleaved floating point audio-data?

With portaudio's  sample-format `paFloat32 | paNonInterleaved` and `Pa_ReadStream()` I'd expect a single array: `&(input_buffer[channel * n_samples])`.

Can you dump the incoming LTC to a file? That might be helpful to check if it's valid LTC or shed some more light on the issue.

**Comment by zhiguang1** [comment]
yes it is non-interleaved floating point audio, i'm parsing the channels out

how would I confirm that this is valid LTC?

**Comment by x42** [comment]
I'd first do a visual check in some sound-file editor.  Expect  2KHz and 4KHz square waves (gapless) like

![ltc](https://cloud.githubusercontent.com/assets/30246/26317904/6fa8ff06-3f19-11e7-8641-f49f801e7a9b.png)

(there are 2 LTC sources overlaid in the image, and the timescale isn't visible, but it should give you a general idea)

If you have a short snippet (1-2 seconds should suffice) upload it and I can have a look.

**Comment by zhiguang1** [comment]
i have an audio snippet but can't upload here, unfortunately.  here is a visual

![image](https://cloud.githubusercontent.com/assets/28867412/26318621/c4edf2d0-3ee9-11e7-9cd8-7d9a7a91c02c.png)

**Comment by x42** [comment]
That looks fine, is that a direct dump of `&data->recordedSamples[6][0]`  or did you record with audacity directly?

**Comment by zhiguang1** [comment]
i recorded with reaper and opened with audacity actually...for a direct dump it would take some more effort

**Comment by x42** [comment]
or try vice versa: test the decoder with some raw data (dumped from the .wav, IIRC audacity can export raw audio data)

**Comment by zhiguang1** [comment]
i'm thinking now that this could be a floating point or float32 issue i.e. exponent/mantissa...will confirm shortly.

**Comment by zhiguang1** [comment]
no that wasn't the issue, i'm not sure if this is of any consequence or use, but it seems to read some data into some of the decoder vars

![image](https://cloud.githubusercontent.com/assets/28867412/26320441/e8c35e0a-3ef0-11e7-8bd2-83cb222fd469.png)

**Comment by zhiguang1** [comment]
so i used RAW audio data from audacity and it is giving such output:

![image](https://cloud.githubusercontent.com/assets/28867412/26323566/5c50056a-3efd-11e7-8fcb-418713f29a7e.png)

**Comment by x42** [comment]
Well, that means LTC without date-information is detected. So it looks like a step forward and it indicates that the issue is indeed with interpreting the portaudio buffer as float array.

It's still odd that the same timecode is parsed over and over again.
What compiler do you use and what's the target platform?  Does `make check` succeed the self-test?

The huge numbers at the end are the correspond to the index passed as last argument to `ltc_decoder_write_float()` for the given decoded LTC frame.

**Comment by zhiguang1** [comment]
if I build in linux, make check passes the self test, but still same issue (decoding to zero) using a contiguous array
